freqtrade/leverage/liquidation_price.py

```diff
@@ -0,0 +1,236 @@
+from typing import Optional
+
+from freqtrade.enums import Collateral, TradingMode, MarginMode
+from freqtrade.exceptions import OperationalException
+
+
+def liquidation_price(
+    exchange_name: str,
+    open_rate: float,
+    is_short: bool,
+    leverage: float,
+    trading_mode: TradingMode,
+    collateral: Optional[Collateral],
+    margin_mode: Optional[MarginMode]
+) -> Optional[float]:
+    if trading_mode == TradingMode.SPOT:
+        return None
+
+    if not collateral:
+        raise OperationalException(
+            "Parameter collateral is required by liquidation_price when trading_mode is "
+            f"{trading_mode}"
+        )
+
+    if exchange_name.lower() == "binance":
+        if not margin_mode:
+            raise OperationalException(
+                f"Parameter margin_mode is required by liquidation_price when exchange is {trading_mode}")
+
+        return binance(open_rate, is_short, leverage, margin_mode, trading_mode, collateral)
+    elif exchange_name.lower() == "kraken":
+        return kraken(open_rate, is_short, leverage, trading_mode, collateral)
+    elif exchange_name.lower() == "ftx":
+        return ftx(open_rate, is_short, leverage, trading_mode, collateral)
+    raise OperationalException(
+        f"liquidation_price is not implemented for {exchange_name}"
+    )
+
+
+def exception(
+    exchange: str,
+    trading_mode: TradingMode,
+    collateral: Collateral,
+    margin_mode: Optional[MarginMode] = None
+):
+    """
+        Raises an exception if exchange used doesn't support desired leverage mode
+        :param exchange: Name of the exchange
+        :param margin_mode: one-way or hedge
+        :param trading_mode: spot, margin, futures
+        :param collateral: cross, isolated
+    """
+    if not margin_mode:
+        raise OperationalException(
+            f"{exchange} does not support {collateral.value} {trading_mode.value} trading ")
+
+    raise OperationalException(
+        f"{exchange} does not support {collateral.value} {margin_mode.value} Mode {trading_mode.value} trading ")
+
+
+def binance(
+    open_rate: float,
+    is_short: bool,
+    leverage: float,
+    margin_mode: MarginMode,
+    trading_mode: TradingMode,
+    collateral: Collateral,
+    **kwargs
```

Comment:
That would make it messy, @xmatthias do you also want how @samgermain is suggesting

Reply:
https://github.com/freqtrade/freqtrade/pull/5379#discussion_r684897541

Reply:
Ok, Thanks for the clarification @samgermain I will make the changes.

Reply:
You will probably need to use `Optional` from the `typing` library

Reply:
Yes i do ...
using kwargs for a method where inputs are clear (and either mandatory, or have predefined defaults) will prevent developer tooling (mypy, ...) from helping us detect errors (like a wrong type, misspelled argument, ...).
You'll have to list the arguments anywhere where you call it this function - so it's better to allow tools to help us detect errors (i'm lazy sometimes, and definitely too lazy to check calls for typos in arguments ...).

If you have 10ds of parameters for a function (like in this case) , maybe consider creating a structure for it (like a NamedTuple or similar) - which will still preserve some safety - while not causing pages-long definitions.

---

I'm however not sure where all these arguments (the data inside) should come from ... they are either part of (one or more) responses from the exchange - or part of the open trade.
Whatever it is, it might make sense to pass that prat into this method ... ? 

If it requires 2-3 calls for binance, and 1 for ftx (different modes to calculate require different data ...) - maybe consider moving the whole logic to the respective exchange classes - where you can "gather" the data from exchange directly without a lot of boilerplate.

Reply:
> I'm however not sure where all these arguments (the data inside) should come from ... they are either part of (one or more) responses from the exchange - or part of the open trade.
Whatever it is, it might make sense to pass that prat into this method ... ?

> If it requires 2-3 calls for binance, and 1 for ftx (different modes to calculate require different data ...) - maybe consider moving the whole logic to the respective exchange classes - where you can "gather" the data from exchange directly without a lot of boilerplate.


@samgermain your opinion? Since you came up with the base structure. I am only filling the gaps here.

Reply:
Why would it require 2-3 calls for binance? Just get whatever data is required using whatever technique you want. Using a NamedTuple sounds like a good idea, or just a class with only properties, I don't know if those are the same thing. @xmatthias do you know what a class with only properties and no methods is in Python?

In general you can just listen to matthias without consulting me.

Reply:
it's not the same but similar ... `@dataclass` sounds good (had to double-check when it was introduced, as we're still supporting python 3.7 - but looks like we're fine (otherwise CI will let us know).